### ISSUE TITLE
Material UI Redesign

### ISSUE THREAD

**Issue by BioHaZard1** [rovo89/XposedInstaller#292]
Hi @rovo89, I had some free time so I thought I'd redesign/materialize the XposedInstaller. Once I'm finished I'll submit a pull request then you can choose if you want to pull or not. I've added some screenshots below so you can see where I'm at with the redesign.

I read before that you're not really a fan of the navigation draw, not sure if this is still the case. However, I could add an option to either show the navigation drawer or the toolbar spinner. Everyone has their own preference I guess. 

![screenshot_2015-08-01-18-07-40](https://cloud.githubusercontent.com/assets/5918419/9022780/c4fdd7d0-387a-11e5-9fd7-6acc28d8eb74.png)
![screenshot_2015-08-01-18-07-43](https://cloud.githubusercontent.com/assets/5918419/9022783/c4ffb7da-387a-11e5-8cd8-f547d0dd7cb4.png)
![screenshot_2015-08-01-18-07-53](https://cloud.githubusercontent.com/assets/5918419/9022781/c4fe5ee4-387a-11e5-8008-fa9f4ca58055.png)
![screenshot_2015-08-01-18-07-57](https://cloud.githubusercontent.com/assets/5918419/9022782/c4ff20cc-387a-11e5-8598-9d52c09d3aab.png)

**Comment by wbedard** [comment]
FWIW...I like it.  It's as simple and clean as the original and, of course, sports the updated look of other apps on Android 5.x.  Since the installer would presumably be used for Android 4.x installs, too, I've heard that app developers have had to jump through various hoops to "backport" the MD look to the older versions of Android.  As long as you've addressed that concern, the only thing that many users might want to see added is a dark/black theme.  It certainly looks like a good effort so far!

**Comment by BioHaZard1** [comment]
I've got all themes working now.

![screenshot 3 aug 2015 18_09_36](https://cloud.githubusercontent.com/assets/5918419/9043206/c70f2370-3a0b-11e5-95bd-f2274e6771f6.png)

**Comment by rovo89** [comment]
Looks good to me. :) Also the navigation drawer looks nice, we could get rid of the dropdown completely (no option required).

I'll be happy to merge it once you're ready, although it might take me some time. I always try to make sure to review every single line of the changes to ensure that it fits with regard to code quality, compatibility and style. So if you have the chance to split your changes up into several commits that touch only certain areas, please do that. Also please follow the existing code style, e.g. no trailing whitespace, correct indention etc. If all of these things are good right from the start, it's much easier to merge than if I have to make those adjustments myself.

Oh, and do you also test this on older Android versions? We don't necessarily need material design there, but it should at least be functional.

**Comment by BioHaZard1** [comment]
Awesome, well I've moved everything over to AppCompat. From what I'm aware that supports everything down to API 7. I'll do some testing on older devices, everything should work fine. I generally create a commit once I change, fix or add something so it should be nice and easy to merge or refactor if necessary. 

Edit: Just compiled a build using API 7, no build errors, everything seems good so far.

**Comment by GermainZ** [comment]
Shouldn't the icons and text have the same color (white would be best IMO)? I'm
aware some Google apps do it this way, but AFAIK the guidelines seem to indicate
the contrary which is more consistent and makes more sense to me. Love it
otherwise.

On Mon, Aug 03, 2015 at 8:16 PM, BioHaZard1 < notifications@github.com [notifications@github.com] > wrote:
I've got all themes working now.

screenshot 3 aug 2015 18_09_36
[https://cloud.githubusercontent.com/assets/5918419/9043206/c70f2370-3a0b-11e5-95bd-f2274e6771f6.png]
[https://cloud.githubusercontent.com/assets/5918419/9043206/c70f2370-3a0b-11e5-95bd-f2274e6771f6.png]

—
Reply to this email directly or view it on GitHub
[https://github.com/rovo89/XposedInstaller/issues/292#issuecomment-127335384] .[https://github.com/notifications/beacon/ACxRlUmEtBBMfaj7naAA2LiiXee_ZUw2ks5oj5mKgaJpZM4FkADb.gif]

**Comment by BioHaZard1** [comment]
Yeah, I've changed that, I personally prefer white too.

**Comment by BioHaZard1** [comment]
Right, I think everything is done now. Just got to make sure all the images resources are present for the different screen densities and I'll create a pull request. Unless you just want to pull the commits from my fork?

**Comment by rovo89** [comment]
I think I'll pick them one by one from your repository. Thanks! Please give me some time though, there are a few more bugs that I'm currently looking at.

**Comment by BioHaZard1** [comment]
Just pushed everything to my repository, I've tried to follow your coding style as must as I can. I haven't changed any of your 'util' classes, they should remain the same. I've pushed all of the code to a new branch called 'material'. There is no rush, take as much time as you need.

https://github.com/BioHaZard1/XposedInstaller/tree/material

Screens of the final version:
![navigation](https://cloud.githubusercontent.com/assets/5918419/9282861/2f4b92e8-42c7-11e5-8439-04700f5a6de4.png)

![installer](https://cloud.githubusercontent.com/assets/5918419/9282865/3a9b5214-42c7-11e5-9947-d5f764f49891.png)

![modules](https://cloud.githubusercontent.com/assets/5918419/9282872/4348854e-42c7-11e5-8120-ced16a3a8ca8.png)

![downloads](https://cloud.githubusercontent.com/assets/5918419/9282876/46fc2bf0-42c7-11e5-91be-f468cf74dabe.png)

![settings](https://cloud.githubusercontent.com/assets/5918419/9282879/4b3ec9ac-42c7-11e5-83d4-94a2b67b91f9.png)

**Comment by InnerFire** [comment]
Its really awesome..I've just test on my phone as well I posted a version of it with additional commits by DVDAndroid in forums of my device.. Many users used it and loved it also..

And I've also got some reviews and reports..

1) Some colors( the purple one in navigation drawer especially) are not proper in Dark theme( its not visible)

2) The navigation drawer is laggy( I have tested it on my Android one 1st Gen having 1 GB RAM)

For first one I've fixed on my form.
This second one is major one but it can be solved using the spinner loading dialog just like play store so that navigation drawer just complete transition without having to wait for fragment to be loaded.


There are few more things about it and I'll consider to report you as  soon as I find more to make sure that this merge will be Perfect for being official.

**Comment by InnerFire** [comment]
@rovo89 I've fixed ( thanks to DVDAndroid) almost all issues as users reported( 3k to 5k users used it on my device's local forum) and also added some features ( swipe to refresh, colorchooser etc again thanks to DVDAndroid) The only problem they have now is translation. So hope you merge it ASAP. All commits are on my local fork. And again it's well tested and approved by users( They were actually awaiting this( and your approval) desperately).

**Comment by rovo89** [comment]
Phew. I finally wanted to start merging this, but I'm already stuck when looking at the huge amount of changes. I guess I'll use your changes as template and copy parts of them manually.

**Comment by BioHaZard1** [comment]
Yeah, that's probably the best move. I have a few other commits containing gradle updates and a few layout fixes improvements.

**Comment by rovo89** [comment]
A few questions I have so far:
- You have replaced some fragments by activities. I'm wondering whether it would make sense to do this for all fragments (except download details). I'm looking at the Google I/O 2014 app as a template as well and this is what they use. Do you see any disadvantages? The only thing I see it that It might be good to keep some fragements for displaying them on one screen for tables. Not sure if this could also be achieved with activities.
- What's the toolbar about? Isn't that simply the action bar?
- Could there be issues due to the missing images for lower resolutions?
- Didn't we have CardViews before?

**Comment by InnerFire** [comment]
For question on resolution, I've also had searched about ( don't exactly remember where I was got the answer but ) the answer was:

Lower resolution are not anymore used much because most of the phones now have big display compared to displays on Old devices that's why it's not anymore used. And some mechanism is also there so that for devices with lower resolutions images will be resized from higher resolutions directly.

**Comment by BioHaZard1** [comment]
- Sure, you could use activities with the navigation drawer however this might somewhat over complicate things. For example, you’d need to have multiple navigation drawers for each activity. Obviously this would require a little more coding but in the end I can’t seen any advantages of doing it this way. In fact, I’ve never seen any app do it this way. All of Google’s other apps do it the same way I did it. 

- Did you mean tablets? if so, do you have a specific way you would like the app to look on tablets.

- A Toolbar is essentially an ActionBar on steroids, it’s classed as view instead of being bound by the applications decor. So this can be placed anywhere inside the interface, you could also have multiple toolbars too. We need it too for the navigation drawer to work correctly. 

- Not really, if the image resource isn’t found/available it will move to the next resource density folder until the resource is found.

- Not really, it was just drawable shaped like a card :). I added the native android cardview, so for devices with API 21 (Lollipop) and higher the shadows are drawn dynamically. If you look on a lighter background you can see the shadow properly. Also a shadow is cast around the whole card to give depth/elevation from the main view.

If there is anything else, or you would like me to explain things better or more in-depth just let me know.

**Comment by rovo89** [comment]
> however this might somewhat over complicate things

Well, in the I/O app, they have added the navdrawer to the base activity that all other activities inherit from. As far as I understood, this would be needed for the settings activity anyway, or how did you solve that?

> All of Google’s other apps do it the same way I did it.

I didn't check in detail, I just noticed that Gmail seems to have different activities as well.

> Note: from what I am aware you can only have one activity on the screen at any one time. So, fragments are the way to go for dynamic layouts.

Yes, that's what I thought. Would definitely be good to stay flexible.

**Comment by BioHaZard1** [comment]
Oh, I see, well you could do it like I did with the settings activity. Essentially, you keep fragment as it is, then you create an activity e.g. SettingsActivity. This adds the settings fragment to a container within the settings activity layout. That way you can launch the fragments as activity via the navigation draw, but as the fragment still exist you can simply add multiple fragments to a single layout for tablets. So it's kind of best of worlds really as you can keep everything dynamic. With XDA One (currently helping with development) I created a floating activity for tablets, this would look good with the DownloadActivity on Xposed.

![screenshot 2 oct 2015 19_42_54](https://cloud.githubusercontent.com/assets/5918419/10254798/db70bbf6-693d-11e5-9b05-35383ea64eb4.png)

**Comment by ArionMiles** [comment]
@BioHaZard1 I'd like to make a design suggestion:

You can drop the left-side pull out navigation bar and utilise a Bottom Bar Navigation they introduced at this year's I/O. Although the material design guidelines say we must use top 3-5 top level destinations in the bottom nav bar, we can have the Framework, Modules, Download, Support, Logs in the bottom bar and you can put a link for the "About" section in the Framework window (since there's a little space on that window which remains unoccupied).

You can avoid using the labels on the bottom nav bar below the icons since it'd be crowded, and only use the icons. 

Here are the guidlines: https://material.google.com/components/bottom-navigation.html
